### PR TITLE
 api: add java proto options to avoid risk of future breakage

### PR DESCRIPTION
**Pull request by dapengzhang0**
*Description*: As discussed with @snowp in https://github.com/envoyproxy/java-control-plane/issues/87#issuecomment-456537248 and https://github.com/envoyproxy/java-control-plane/issues/87#issuecomment-457649859, adding these options will avoid the risk of having API breaking changes on the generated java code as the proto files evolve.

*Risk Level*:  Medium
Note that e.g. [envoyproxy/java-control-plane](https://github.com/envoyproxy/java-control-plane) will have to do a refactoring when they sync the proto files next time.
*Testing*:
*Docs Changes*:
*Release Notes*:
[Optional Fixes #Issue] envoyproxy/java-control-plane#87
[Optional *Deprecated*:]

### CODE REVIEW

File: tools/check_format.py [hunk 1]
```diff
@@ -172,14 +172,64 @@ def fixJavaProtoOptions(file_path):
...
+  for line in fileinput.FileInput(file_path):
+    if "option java_outer_classname = " in line:
+      return []
+
+  return ["Java proto option 'java_outer_classname' not set for file: %s" % file_path]
+
+
+def checkJavaMultipleFilesProtoOption(file_path):
+  for line in fileinput.FileInput(file_path):
+    if "option java_multiple_files = " in line:
```
Comment (snowp) [inline tools/check_format.py]: Maybe check that this is set to true? Or are there cases where you wouldn't want that?

File: tools/check_format.py [hunk 2]
```diff
@@ -172,14 +172,64 @@ def fixJavaProtoOptions(file_path):
...
   for line in fileinput.FileInput(file_path):
     if "option java_package = \"io.envoyproxy.envoy" in line:
       return []
 
   return ["Java proto option 'java_package' not set correctly for file: %s" % file_path]
 
 
+def checkJavaOuterClassnameProtoOption(file_path):
+  for line in fileinput.FileInput(file_path):
+    if "option java_outer_classname = " in line:
```
Comment (snowp) [inline tools/check_format.py]: Should this be checking for the generated name? Does that matter?

File: tools/check_format.py [hunk 3]
```diff
@@ -172,14 +172,64 @@ def fixJavaProtoOptions(file_path):
...
+    if line.startswith("package "):
+      line = line.replace(line, line + to_add)
+    sys.stdout.write(line)
+
+  return []
+
+
+def fixJavaMultipleFilesProtoOption(file_path):
+  for line in fileinput.FileInput(file_path):
+    if "option java_multiple_files = " in line:
```
Comment (snowp) [inline tools/check_format.py]: Same here, should this check for true?

File: tools/check_format.py [hunk 4]
```diff
@@ -172,14 +172,64 @@ def fixJavaProtoOptions(file_path):
...
+def fixJavaOuterClassnameProtoOption(file_path):
+  file_name = os.path.basename(file_path)[:-len(".proto")]
+  if "-" in file_name or "." in file_name or not file_name.islower():
+    return ["Unable to decide java_outer_classname for proto file: %s" % file_path]
+
+  for line in fileinput.FileInput(file_path):
+    if "option java_outer_classname = " in line:
+      return []
+
+  to_add = "option java_outer_classname = \"" + ''.join(
```
Comment (snowp) [inline tools/check_format.py]: nit: maybe split this line into two lines to make it easier to read?

File: tools/check_format.py [hunk 5]
```diff
@@ -172,14 +172,64 @@ def fixJavaProtoOptions(file_path):
   return []
 
 
-def checkJavaProtoOptions(file_path):
+# Add 'option java_outclass_name = FooBarProto' for foo_bar.proto
```
Comment (snowp) [inline tools/check_format.py]: nit: typo outclass -> outer_classname

Comment (dapengzhang0) [reply tools/check_format.py]: Done.

File: tools/check_format.py (hunk 4 above)
Comment (dapengzhang0) [reply tools/check_format.py]: Done.

File: tools/check_format.py (hunk 3 above)
Comment (dapengzhang0) [reply tools/check_format.py]: Same here, allow purposely setting it to false for some particular proto file.

File: tools/check_format.py (hunk 2 above)
Comment (dapengzhang0) [reply tools/check_format.py]: This allows purposely setting custom outer classname for some particular proto file. There's PRO and CON, but may not matter too much yet.

File: tools/check_format.py (hunk 1 above)
Comment (dapengzhang0) [reply tools/check_format.py]: This only checks if the developer forgot to set it. If the developer sets it to false, that is unusual, might imply that they purposely want it to be false.

Comment (snowp) [reply tools/check_format.py]: Ok, makes sense

File: api/envoy/config/accesslog/v2/file.proto [hunk 6]
```diff
@@ -1,6 +1,8 @@
 syntax = "proto3";
 
 package envoy.config.accesslog.v2;
+option java_outer_classname = "FileProto";
```
Comment (htuch) [inline api/envoy/config/accesslog/v2/file.proto]: Nit: I think a blank line between `package` and the `option` block would make sense here and elsewhere.

File: tools/check_format.py [hunk 7]
```diff
@@ -172,14 +172,65 @@ def fixJavaProtoOptions(file_path):
...
+def fixJavaOuterClassnameProtoOption(file_path):
+  file_name = os.path.basename(file_path)[:-len(".proto")]
+  if "-" in file_name or "." in file_name or not file_name.islower():
+    return ["Unable to decide java_outer_classname for proto file: %s" % file_path]
+
+  for line in fileinput.FileInput(file_path):
+    if "option java_outer_classname = " in line:
+      return []
+
+  to_add = "option java_outer_classname = \"" \
```
Comment (htuch) [inline tools/check_format.py]: Nit: mixture of `"` and ` style for Python strings, please stick with the dominate convention in the file.

File: tools/check_format.py [hunk 8]
```diff
@@ -172,14 +172,65 @@ def fixJavaProtoOptions(file_path):
...
 
 
-def checkJavaProtoOptions(file_path):
+# Add 'option java_outer_classname = FooBarProto' for foo_bar.proto
+def fixJavaOuterClassnameProtoOption(file_path):
+  file_name = os.path.basename(file_path)[:-len(".proto")]
+  if "-" in file_name or "." in file_name or not file_name.islower():
+    return ["Unable to decide java_outer_classname for proto file: %s" % file_path]
+
+  for line in fileinput.FileInput(file_path):
```
Comment (htuch) [inline tools/check_format.py]: You don't need to iterate, just read in the file and do an `in` operation.

File: tools/check_format.py [hunk 9]
```diff
@@ -172,14 +172,65 @@ def fixJavaProtoOptions(file_path):
...
+    if "option java_outer_classname = " in line:
+      return []
+
+  to_add = "option java_outer_classname = \"" \
+       + ''.join(x.title() for x in file_name.split('_')) \
+       + "Proto\";\n"
+
+  for line in fileinput.FileInput(file_path, inplace=True):
```
Comment (htuch) [inline tools/check_format.py]: Don't need to iterate, can just do a `re.sub` here or similar on the entire file contents.

File: tools/check_format.py [hunk 10]
```diff
@@ -172,14 +172,65 @@ def fixJavaProtoOptions(file_path):
...
+    if line.startswith("package "):
+      line = line.replace(line, line + to_add)
+    sys.stdout.write(line)
+
+  return []
+
+
+def fixJavaMultipleFilesProtoOption(file_path):
```
Comment (htuch) [inline tools/check_format.py]: This is similar logic to the above function, can just refactor and parameterize on the string.

File: tools/check_format.py (hunk 9 above)
Comment (dapengzhang0) [reply tools/check_format.py]: Done.

File: tools/check_format.py (hunk 10 above)
Comment (dapengzhang0) [reply tools/check_format.py]: Done.

File: tools/check_format.py [hunk 11]
```diff
@@ -147,37 +147,61 @@ def checkNamespace(file_path):
...
+    if substring in text:
       return []
 
-  if package_name is None:
-    return ["Unable to find package name for proto file: %s" % file_path]
+    new_text = re.sub(PROTO_PACKAGE_REGEX, repl, text)
+    f.seek(0)
+    f.write(new_text)
+    f.truncate()
+    sys.stdout.write(new_text)
```
Comment (htuch) [inline tools/check_format.py]: Remove debug

File: tools/check_format.py [hunk 12]
```diff
@@ -147,37 +147,61 @@ def checkNamespace(file_path):
...
+    text = f.read()
+    if substring in text:
       return []
 
-  if package_name is None:
-    return ["Unable to find package name for proto file: %s" % file_path]
+    new_text = re.sub(PROTO_PACKAGE_REGEX, repl, text)
+    f.seek(0)
+    f.write(new_text)
+    f.truncate()
```
Comment (htuch) [inline tools/check_format.py]: Why is this needed?

File: tools/check_format.py [hunk 13]
```diff
@@ -147,37 +147,61 @@ def checkNamespace(file_path):
...
+
+  with open(file_path, "r+") as f:
+    text = f.read()
+    if substring in text:
       return []
 
-  if package_name is None:
-    return ["Unable to find package name for proto file: %s" % file_path]
+    new_text = re.sub(PROTO_PACKAGE_REGEX, repl, text)
+    f.seek(0)
```
Comment (htuch) [inline tools/check_format.py]: I think you could do this as two separate `with` blocks, one with mode `r` and the other with `w`. This would be conceptually cleaner.

File: tools/check_format.py [hunk 14]
```diff
@@ -147,37 +147,61 @@ def checkNamespace(file_path):
...
-  for line in fileinput.FileInput(file_path, inplace=True):
-    if line.startswith("package "):
-      line = line.replace(line, line + to_add)
-    sys.stdout.write(line)
+def packageNameForProto(file_path):
+  package_name = None
+  error_message = []
+  with open(file_path) as f:
+    result = PROTO_PACKAGE_REGEX.search(f.read())
+    if not result is None and len(result.groups()) == 1:
```
Comment (htuch) [inline tools/check_format.py]: `if result is not None and ...`

File: tools/check_format.py [hunk 15]
```diff
@@ -211,6 +235,14 @@ def findSubstringAndReturnError(pattern, file_path, error_message):
     return []
 
 
+def errorIfNoSubstringFound(pattern, file_path, error_message):
+  with open(file_path) as f:
+    if pattern in f.read():
+      return []
+    return [file_path + ": " + error_message]
+    # return [] if pattern in f.read() else [file_path + ": " + error_message]
```
Comment (htuch) [inline tools/check_format.py]: Debug?

File: tools/check_format.py [hunk 16]
```diff
@@ -147,37 +147,61 @@ def checkNamespace(file_path):
...
+                                     to_add)
+
+
+# Add "option java_outer_classname = FooBarProto;" for foo_bar.proto
+def fixJavaOuterClassnameProtoOption(file_path):
+  file_name = os.path.basename(file_path)[:-len(".proto")]
+  if "-" in file_name or "." in file_name or not file_name.islower():
+    return ["Unable to decide java_outer_classname for proto file: %s" % file_path]
+
+  to_add = "option java_outer_classname = \"" \
```
Comment (htuch) [inline tools/check_format.py]: FWIW, I think if you make some of these strings like `option java_outer_classname = \"` file-level constants, it could make it easier to keep the fix/check synced.

File: tools/check_format.py [hunk 17]
```diff
@@ -147,37 +150,58 @@ def checkNamespace(file_path):
...
+def insertProtoOptionIfNotFound(substring, file_path, to_add):
+  text = None
+  with open(file_path) as f:
+    text = f.read()
 
-      package_name = result.group(1)
-    if "option java_package = \"io.envoyproxy.envoy" in line:
-      return []
+  if not substring in text:
+
```
Comment (htuch) [inline tools/check_format.py]: Nit: superfluous blank line.

**Comment by htuch** [comment]
Please merge master to pick up https://github.com/envoyproxy/envoy/pull/5827.
